microceph/ceph/cephfs_volume.go

```diff
@@ -0,0 +1,155 @@
+package ceph
+
+import (
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/canonical/microceph/microceph/logger"
+	"github.com/tidwall/gjson"
+)
+
+type CephFSVolume struct {
+	Name                string
+	SubvolumeGroups     map[string]Subvolumegroup
+	UngroupedSubVolumes []UngroupedSubvolume
+}
+
+type Subvolumegroup struct {
+	SubVolumes []GroupedSubvolume
+}
+
+type (
+	Volume             string
+	Subvolume          string
+	GroupedSubvolume   string
+	UngroupedSubvolume string
+	Subvolumes         []Subvolume
+)
+
+// Example path of a subvolume /volumes/subvolumegroup/subvolume/
+const (
+	CephFsConstrantEmptyIndex         = 0 // ""
+	CephFsConstrantStringVolumesIndex = 1 // "volumes"
+	CephFsSubVolumeGroupIndex         = 2 // "subvolume group name"
+	CephFsSubVolumeIndex              = 3 // "subvolume name"
+)
+
+func ListCephFSVolumes() ([]Volume, error) {
+	args := []string{"fs", "volume", "ls", "--format=json"}
+	output, err := cephRun(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("FSVOL: listed %s as cephfs volume", output)
+
+	volumes := gjson.Get(output, "#.name")
+	response := make([]Volume, 0, len(volumes.Array()))
+	for _, volume := range volumes.Array() {
+		if len(volume.String()) == 0 {
+			continue
+		}
+
+		logger.Infof("FSVOL: found %s as cephfs volume", volume.String())
+		response = append(response, Volume(volume.String()))
+	}
+
+	return response, nil
+}
+
+func CephFsSubvolumePathDeconstruct(path string) (subvolumegroup string, subvolume string, err error) {
+	parts := strings.Split(path, "/")
+	if len(parts) < 3 {
+		return "", "", fmt.Errorf("invalid CephFS subvolume path: %s", path)
+	}
+
+	subvolumegroup = parts[CephFsSubVolumeGroupIndex]
+	subvolume = parts[CephFsSubVolumeIndex]
+
+	return subvolumegroup, subvolume, nil
+}
+
+func GetCephFSVolume(volume Volume) (CephFSVolume, error) {
+	response := CephFSVolume{Name: string(volume)}
+	var err error
+
+	response.SubvolumeGroups, err = GetCephFSSubvolumeGroups(volume)
+	if err != nil {
+		return response, fmt.Errorf("failed to get subvolume groups for CephFS volume %s: %w", volume, err)
+	}
+
+	Subvolumes, err := GetCephFSSubvolumes(volume, "")
+	if err != nil {
+		return response, fmt.Errorf("failed to get ungrouped subvolumes for CephFS volume %s: %w", volume, err)
+	}
+
+	response.UngroupedSubVolumes = unsafe.Slice(
+		(*UngroupedSubvolume)(unsafe.SliceData(Subvolumes)),
+		len(Subvolumes),
+	)
```

Comment:
Bit uneasy about the unsafety, could we do sth like the below instead?

```
response.UngroupedSubVolumes = make([]UngroupedSubvolume, len(Subvolumes))
for i, sv := range Subvolumes {
    response.UngroupedSubVolumes[i] = UngroupedSubvolume(sv)
}
```

Similarly below

Reply:
Ack, gotcha. I view this as a tradeoff between performance / robustness -- with the unsafe construct we give up some static type checking which could impact future refactorings etc.

Is this an operation that you expect be run repeatedly / frequently? I'd advise to only go for the unsafe construct if we expect real world impact.

Reply:
fair enough, I have modified it to follow the safer copy construct and prepare the slice.